### PR TITLE
executor: show better placement schedule state

### LINKED ISSUE #30960: placement: show a better hint message for schduleing state
## Enhancement

> The replication API should report "UNABLE TO SCHEDULE" instead of "INPROGRESS" if the rule can not be applied immediately. There are many such scenarios where this could happen (followers=N and stores is < N).

Refer tikv/pd#4467.

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Issue Number: close #30960 

Problem Summary: Depends on tikv/pd#4508. Now we have "PENDING" to indicate PD did not start the schedule at all. It is possibly caused by busy looping or unmet condition, whatever, it is not scheduling.

I'll complete the document in another PR, since there are other document changes.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
```
mysql root@127.0.0.1:(none)> create table test.t1 (a int) FOLLOWERS=4;
Query OK, 0 rows affected
Time: 0.200s
mysql root@127.0.0.1:(none)> show placement;
+---------------+-------------+------------------+
| Target        | Placement   | Scheduling_State |
+---------------+-------------+------------------+
| TABLE test.t1 | FOLLOWERS=4 | PENDING       |
+---------------+-------------+------------------+
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: executor/show_placement.go [hunk 1]
```diff
@@ -421,39 +442,55 @@ func (e *ShowExec) getPolicyPlacement(policyRef *model.PolicyRefInfo) (settings
...
+func fetchScheduleState(ctx context.Context, scheduleState map[int64]placementScheduleState, id int64) (placementScheduleState, error) {
 	if s, ok := scheduleState[id]; ok {
 		return s, nil
 	}
 	startKey := codec.EncodeBytes(nil, tablecodec.GenTablePrefix(id))
 	endKey := codec.EncodeBytes(nil, tablecodec.GenTablePrefix(id+1))
 	schedule, err := infosync.GetReplicationState(ctx, startKey, endKey)
+	var state placementScheduleState
+	switch schedule {
+	case "\"REPLICATED\"\n":
```
Comment (lcwangchao) [inline executor/show_placement.go:454]: It's strange...

Comment (xhebox) [reply executor/show_placement.go:454]: This is the only entry, it is overkill to parse the json. It is certain that there are only three possibilities.

Comment (xhebox) [reply executor/show_placement.go:454]: But yeah, maybe deserialize it in `GetReplicationState`.

Comment (lcwangchao) [reply executor/show_placement.go:454]: > This is the only entry, it is overkill to parse the json. It is certain that there are only three possibilities.

I don't think so... Without JSON parser, we must make an assumption that the return value equal to `"REPLICATED"\n` . But `"REPLICATED"\n\n` and `"REPLICATED"` are still legal values and the API maintainer may not be aware it will cause a break when some day it changes to the above format. 

In other words, it is safer to use a JSON parser than string compare

Comment (xhebox) [reply executor/show_placement.go:454]: I believe the API maintainer is me. And pretty much every JSON marshal of golang, will results into `"..."` for a string. Then PD will add another '\n' automatically, which has been around for years. If you really insist so, `StartsWith` is enough to ignore the unknown suffix.

**Comment by xhebox** [comment]
/cc @morgo 

PTAL, not about the code logic, but the name of third state. Currently I made it "WAITING". This primarily comes from the truth that it is not equal to "UNABLE TO SCHEDULE".

PD could only tell that it is not scheduling, i.e. it paused the schedule somehow due to other conditions. PD could recover from "WAITING" to "INPROGRESS"/"SCHEDULED" at anytime, or vice versa.

Reviewers of PD also said about "WAITING" and "INPROGRESS" is too similar. But "unable to schedule" feels like it won't schedule at all. What about 'NOT SCHEDULING'? I really ran out of words in my mind.

**Comment by morgo** [comment]
> PTAL, not about the code logic, but the name of third state. Currently I made it "WAITING". This primarily comes from the truth that it is not equal to "UNABLE TO SCHEDULE".

I see. I think [PENDING](https://www.merriam-webster.com/dictionary/pending) is the correct word here. It is common to say that "the case is pending out a decision".

**Comment by xhebox** [comment]
> > PTAL, not about the code logic, but the name of third state. Currently I made it "WAITING". This primarily comes from the truth that it is not equal to "UNABLE TO SCHEDULE".
> 
> I see. I think [PENDING](https://www.merriam-webster.com/dictionary/pending) is the correct word here. It is common to say that "the case is pending out a decision".

Neat!

**Comment by morgo** [comment]
@xhebox great job btw, thanks for working on this!

File: domain/infosync/region.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	switch t {
+	case PlacementScheduleStateScheduled:
+		return "SCHEDULED"
+	case PlacementScheduleStateInProgress:
+		return "INPROGRESS"
+	case PlacementScheduleStatePending:
+		return "PENDING"
+	default:
+		return "WAITING"
+	}
```
Comment (morgo) [inline domain/infosync/region.go:49]: Is this going to be confusing because 'WAITING' could still be returned?

File: domain/infosync/region.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		// it should not fail
+		var state string
+		_ = json.Unmarshal(res, &state)
+		switch state {
+		case "REPLICATED":
+			st = PlacementScheduleStateScheduled
+		case "INPROGRESS":
+			st = PlacementScheduleStateInProgress
+		case "WAITING":
+			st = PlacementScheduleStatePending
```
Comment (morgo) [inline domain/infosync/region.go:81]: I didn't see this when I commented that 'PENDING' is better than 'WAITING'. If we are returning 'WAITING' from the HTTP API but 'PENDING' from SQL, it could lead to confusion :(

Are we able to change it in pd as well, or do we have to change back to "WAITING" for uniformity? :(

Comment (xhebox) [reply domain/infosync/region.go]: It is just a mistake. Yeah, it has been fixed.

**Review by lcwangchao — APPROVED** [review]

File: domain/infosync/region.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	if len(addrs) == 0 {
+		return PlacementScheduleStatePending, errors.Errorf("pd unavailable")
+	}
+
+	res, err := doRequest(ctx, addrs, fmt.Sprintf("%s/replicated?startKey=%s&endKey=%s", pdapi.Regions, hex.EncodeToString(startKey), hex.EncodeToString(endKey)), "GET", nil)
+	if err == nil && res != nil {
+		st := PlacementScheduleStatePending
+		// it should not fail
+		var state string
+		_ = json.Unmarshal(res, &state)
```
Comment (hawkingrei) [inline domain/infosync/region.go:74]: Why is the ```error``` of ```json.Unmarshal``` ignored?

Comment (hawkingrei) [reply domain/infosync/region.go:74]: we should defend against the error and not think about whether it fails.

File: executor/show_placement.go [hunk 5]
```diff
@@ -462,26 +466,25 @@ func fetchTableScheduled(ctx context.Context, scheduleState map[int64]bool, tabl
...
-func toScheduleStateString(schedule bool) string {
-	if schedule {
-		return "SCHEDULED"
+func accumulateState(curr, news infosync.PlacementScheduleState) infosync.PlacementScheduleState {
+	a, b := int(curr), int(news)
+	if a > b {
+		return news
 	}
-	return "INPROGRESS"
+	return curr
```
Comment (hawkingrei) [inline executor/show_placement.go:489]:
```
return math.min(int(curr), int(news))
```

File: domain/infosync/region.go (hunk 4 above)
Comment (hawkingrei) [reply domain/infosync/region.go:74]: The maximum length of JSON strings will be checked here.

Comment (xhebox) [reply domain/infosync/region.go:74]: The result from PD just can not be too long. It really should not give error.

And assuming that `unmarshal` fails, we would just return `PENDING` as the default fallback, which totally makes sense for me. If you check it, you could return a different word like `ERROR` though.

**Review by hawkingrei — APPROVED** [review]
